### PR TITLE
update: lost pixel plugin for visual regression testing

### PR DESCRIPTION
**Pull request by nilubava**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,27 +1,47 @@
...
+        run: npx gatsby build --prefix-paths
+        env:
+          GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES: true
+          CI: true
+
+      - name: Run Gatsby app
+        run: npx gatsby serve --host 172.17.0.1 &
+
+      - name: Lost Pixel
+        uses: lost-pixel/lost-pixel@465da38bb5641745fbea0fb5f98377bd3367be21
```
Comment (janpio) [inline .github/workflows/ci.yml]: Unnecessary pin for a specific commit I think

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,27 +1,46 @@
...
+
+      - name: Lost Pixel
+        uses: lost-pixel/lost-pixel@next
+        env:
+          LOST_PIXEL_API_KEY: clb5felu61843101qqg5entz4x
+
+      - uses: actions/upload-artifact@v3
+        with:
+          name: lost-pixel-artefacts
+          path: .lostpixel
```
Comment (janpio) [inline .github/workflows/ci.yml]: Not sure if that still exists or would be useful - let's see.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,27 +1,46 @@
...
+on: [push]
 
 jobs:
   gatsby-for-percy:
-    name: Percy screenshots via Gatsby build
+    name: Lost Pixel screenshots via Gatsby build
     runs-on: ubuntu-latest
     env:
-      PERCY_TOKEN: ${{ secrets.PERCY_TOKEN }}
       FONTAWESOME_NPM_AUTH_TOKEN: ${{ secrets.FONTAWESOME_NPM_AUTH_TOKEN }}
```
Comment (janpio) [inline .github/workflows/ci.yml]: Doesn't this also need to be set via `npm config` to work for Gatsby?

File: lostpixel.config.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        name: '-reference-api-reference-prisma-client-reference',
+      },
+    ],
+    // you create your pages here & both are composed
+    // pagesJsonUrl: 'http://localhost:9000/lost-pixel.json',
+    // baseUrl: 'http://localhost:9000',
+    // pagesJsonUrl: 'http://172.17.0.1:9000/lost-pixel.json',
+    baseUrl: 'http://172.17.0.1:9000',
+  },
+  lostPixelProjectId: 'clb5felu61843101qqg5entz4x',
```
Comment (janpio) [inline lostpixel.config.js]:
```suggestion
  lostPixelProjectId: 'clb5ek3mm1772001qqg7yban38',
```

File: lostpixel.config.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  commitHash: process.env.COMMIT_HASH,
+  compareEngine: 'odiff',
+  //   s3: {
+  //     endPoint: 'ams3.digitaloceanspaces.com',
+  //     bucketName: 'beta-seven-turtle-time-blue-narrow',
+  //     accessKey: process.env.S3_ACCESS_KEY,
+  //     secretKey: process.env.S3_SECRET_KEY,
+  //   },
+  compareConcurrency: 10,
+  shotConcurrency: 1,
```
Comment (janpio) [inline lostpixel.config.js]: All these are probably optional and automated and can be commented out or removed I would guess.

File: lostpixel.config.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  beforeScreenshot: async (page) => {
+    await page.addStyleTag({
+      content: `iframe {
+        visibility: hidden;
+      }
+  
+      /* do not show underline animation */
+      #toc-holder  a {
+        background-size: 0 !important;
```
Comment (d-ivashchuk) [inline lostpixel.config.js]:
```suggestion
module.exports = {
  pageShots: {
    // write your pages here, it's an empty array, as we are generating the pages on the fly in gatsby-plugin
    pages: [],
    // you create your pages on the fly here & it is composed with `pages` array
    pagesJsonUrl: 'http://localhost:9000/lost-pixel.json',
  },
  lostPixelProjectId: 'clb5ek3mm1772001qqg7yban38',
  apiKey: process.env.LOST_PIXEL_API_KEY,
  compareConcurrency: 10,
  shotConcurrency: 1,
  beforeScreenshot: async (page) => {
    await page.addStyleTag({
      content: `iframe {
        visibility: hidden;
      }
  
      /* do not show underline animation */
      #toc-holder  a {
        background-size: 0 !important;
```

File: plugins/gatsby-plugin-lost-pixel/gatsby-node.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  }`
+  const { data } = await graphql(query)
+
+  // Construct the pages json by iterating over the mdx files.
+
+  console.log({ edges: data.allSitePage.edges })
+  const pages = data.allSitePage.edges
+    .map((edge, i) => {
+      // Skip explicitly excluded paths
+      if (excludedPaths.includes(edge.node.path)) return null
```
Comment (d-ivashchuk) [inline plugins/gatsby-plugin-lost-pixel/gatsby-node.js]:
```suggestion
      if (excludedPaths.includes(edge.node.path) || longPages.includes(edge.node.path)) return null
```

File: plugins/gatsby-plugin-lost-pixel/gatsby-node.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      // Allow headless browser to render super long pages before screenshoting them
+
+      if (longPages.includes(edge.node.path)) {
+        console.log('Landed into long page, increasing wait time before screenshot')
+        return {
+          path: edge.node.path,
+          name: edge.node.path.split('/').join('-'),
+          waitBeforeScreenshot: 10000,
+        }
+      }
```
Comment (d-ivashchuk) [inline plugins/gatsby-plugin-lost-pixel/gatsby-node.js]:
```suggestion

```

File: plugins/gatsby-plugin-lost-pixel/gatsby-node.js (hunk 7 above)
Comment (d-ivashchuk) [reply plugins/gatsby-plugin-lost-pixel/gatsby-node.js]: let's not render the long pages at all for now and get it running. We will figure them out when everything else is stable & working.

File: lostpixel.config.js (hunk 6 above)
Comment (d-ivashchuk) [reply lostpixel.config.js]: `pagesJsonUrl` is just a hosted JSON that has the same format as `pages: [Page]`, we just create that programatically and consume on line `6`

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -1,27 +1,46 @@
...
+        run: npx gatsby build --prefix-paths
+        env:
+          GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES: true
+          CI: true
+
+      - name: Run Gatsby app
+        run: npx gatsby serve --host 172.17.0.1 &
+
+      - name: Lost Pixel
+        uses: lost-pixel/lost-pixel@next
```
Comment (d-ivashchuk) [inline .github/workflows/ci.yml]:
```suggestion
        uses: lost-pixel/lost-pixel@v3.0.4
```

**Review by d-ivashchuk — APPROVED** [review]

**Pull request merged**